modules/python/pylib/syslogng/dest.py

```diff
@@ -0,0 +1,97 @@
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# As an additional exemption you are allowed to compile & link against the
+# OpenSSL libraries as published by the OpenSSL project. See the file
+# COPYING for details.
+#
+#############################################################################
+try:
+    from _syslogng import LogDestination
+
+except ImportError:
+    import warnings
+    warnings.warn("You have imported the syslogng package outside of syslog-ng, thus some of the functionality is not available. Defining fake classes for those exported by the underlying syslog-ng code")
+
+    LogDestination = object
+
+class LogDestination(LogDestination):
+    LTR_DROP = 0
+    LTR_ERROR = 1
+    LTR_EXPLICIT_ACK_MGMT = 2
+    LTR_SUCCESS = 3
+    LTR_QUEUED = 4
+    LTR_NOT_CONNECTED = 5
+    LTR_RETRY = 6
+
+    def open(self):
+        """Open a connection to the target service"""
+        return True
+
+    def close(self):
+        """Close the connection to the target service"""
+        pass
+
+    def is_opened(self):
+        """Check if the connection to the target is able to receive messages"""
+        return True
+
+    def init(self, options):
+        """Initialize this log element
+
+        This method is called when the syslog-ng configuration is
+        initialized.  You can prevent the configuration to be accepted by
+        returning False, in which case you should log why that was the case.
+
+        The same object can be initalized multiple times in case the
+        configuration reload operation fails in syslog-ng:
+          1) init() when the config is first initialized
+          2) deinit() as a preparation for reload
+          3) initialization of the new config fails for some reason (any one
+             object can cause it to fail)
+          4) init() again, as syslog-ng reverts to the old config in such cases.
+
+        Args:
+            options:
+                is a dictionary that contains options passed using the
+                options() parameter of the python() destination
+
+        Returns:
+            bool: True if successful, False otherwise
+        """
+        return True
+
+    def deinit(self):
+        """Deinitialize this log element
+
+        This method is called when the syslog-ng configuration is being shut
+        down.
+        """
+        pass
+
+    def send(self, msg):
```

Comment:
Defining these methods for documenting their usage in code is a good idea, but you also defined mandatory methods that do nothing.

Since our intention is to inherit from these classes, we lose a lot of validation:
`python-dest: Error initializing Python destination, class does not have a send() method`, etc.

Reply:
this is true, maybe we should raise a NotImplementedError instead.

I don't think that the validation from the C code would trump the fact that these are now clearly visible methods with a nice description. We could even generate documentation for the library from sphinx (and we should)

I am adding the NotImplementedError exceptions with the current iteration, let me know what you think about this.

Reply:
`NotImplementedError` sounds much better in my opinion. The only part I'm concerned about is the good ol' syslog-ng way of validating things: we report almost all the persistent errors before init() and we shut down syslog-ng once we know we won't work properly.

`NotImplementedError` is too late and syslog-ng can not be stopped, for example, after the first non-implemented send() call.

The above reasoning can be invalidated easily: this is a programming error, not a configuration issue.

I'm going to have a final look at this tomorrow and approve/merge the PR.

Thank you for thinking this through.